ml-algorithms/src/main/java/org/opensearch/ml/engine/tools/CatIndexTool.java

```diff
@@ -131,57 +144,150 @@ public <T> void run(Map<String, String> parameters, ActionListener<T> listener)
+                // 1) the deleted index was explicitly passed as parameter to the /_cat/indices request. In this case we
+                // want the subsequent requests to fail.
+                // 2) the deleted index was resolved as part of a wildcard or _all. In this case, we want the subsequent
+                // requests not to fail on the deleted index (as we want to ignore wildcards that cannot be resolved).
+                // This behavior can be ensured by letting the cluster state, cluster health and indices stats requests
+                // re-resolve the index names with the same indices options that we used for the initial cluster state
+                // request (strictExpand).
+                sendClusterStateRequest(indices, subRequestIndicesOptions, local, client, new ActionListener<>() {
+                    @Override
+                    public void onResponse(ClusterStateResponse clusterStateResponse) {
+                        // Starts to fetch index details here, if a batch fails build whatever we have and return.
+                        fetchPages(
+                            indices,
+                            local,
+                            DEFAULT_CLUSTER_MANAGER_NODE_TIMEOUT,
+                            includeUnloadedSegments,
+                            pageParams,
+                            pageResults,
+                            clusterStateResponse,
+                            getSettingsResponse,
+                            subRequestIndicesOptions,
+                            originalListener
+                        );
+                    }
+
+                    @Override
+                    public void onFailure(final Exception e) {
+                        originalListener.onFailure(e);
+                    }
+                });
+            }
+
+            @Override
+            public void onFailure(final Exception e) {
+                originalListener.onFailure(e);
+            }
+        });
+    }
+
+    private void fetchPages(
+        String[] indices,
+        boolean local,
+        TimeValue clusterManagerNodeTimeout,
+        boolean includeUnloadedSegments,
+        PageParams pageParams,
+        Queue<Collection<ActionResponse>> pageResults,
+        ClusterStateResponse clusterStateResponse,
+        GetSettingsResponse getSettingsResponse,
+        IndicesOptions subRequestIndicesOptions,
+        ActionListener<Table> originalListener
+    ) {
+        final ActionListener<PageToken> iterativeListener = ActionListener.wrap(r -> {
+            // when previous response returns, build next request with response and invoke again.
+            PageParams nextPageParams = new PageParams(r.getNextToken(), pageParams.getSort(), pageParams.getSize());
+            // when next page doesn't exist or reaches max supported page size, return.
+            if (r.getNextToken() == null || pageResults.size() >= MAX_SUPPORTED_LIST_INDICES_PAGE_SIZE) {
```

Comment:
I may be missing something here but this comment is confusing.

I get that if the next page doesn't exist, we can return, but why return even if the next token is not null but we've hit the max?  This seems like we'd never hit the next page (since we would only have a next token in the case this page hit the max and had to paginate).

Also it seems the "max" could fit in one page.  Should this be `>` rather than `>=` here?  Or will we never get `>` and it could be `==` and we're just coding defensively?

This seems not right.  I don't see a unit or IT test at the edge case of `MAX_SUPPORTED_LIST_INDICES_PAGE_SIZE` which is really necessary to properly test this.

Reply:
This is from OS [core](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/rest/action/list/RestIndicesListAction.java#L67-L69), it's limiting the max page size to the value of `MAX_SUPPORTED_LIST_INDICES_PAGE_SIZE` means you can get at most this number of page results, so here I'm using `>=`, this can be changed to `==` but using `>=` for defensive.

Reply:
OK, now that I understand the purpose of this constant, this means you may have reached many pages but even if there are more past this max, you're going to stop.